### PR TITLE
ARROW-11425: [C++][Compute] Optimize quantile kernel for integers

### PR DESCRIPTION
**Pull request by cyb70289**
For integers within limited value range, a histogram approach is useful
to reduce memory footprint and improve performance.

### CODE REVIEW

**Comment by cyb70289** [comment]
Tested on Xeon Gold 5218, clang-9.
Up to 2x improvements for integers with limited value range, minor drop for 50% nulls.

Before
```
QuantileKernelInt32/1048576/10000        2058 us         2058 us          340 bytes_per_second=485.836M/s null_percent=0.01 size=1048.58k
QuantileKernelInt32/1048576/100          1279 us         1279 us          551 bytes_per_second=781.752M/s null_percent=1 size=1048.58k
QuantileKernelInt32/1048576/10           1719 us         1719 us          408 bytes_per_second=581.835M/s null_percent=10 size=1048.58k
QuantileKernelInt32/1048576/2            1631 us         1631 us          429 bytes_per_second=613.259M/s null_percent=50 size=1048.58k
QuantileKernelInt32/1048576/1           0.601 us        0.601 us      1164926 bytes_per_second=1.58694T/s null_percent=100 size=1048.58k
QuantileKernelInt32/1048576/0            1864 us         1864 us          376 bytes_per_second=536.431M/s null_percent=0 size=1048.58k

QuantileKernelInt64/1048576/10000         944 us          944 us          741 bytes_per_second=1059.49M/s null_percent=0.01 size=1048.58k
QuantileKernelInt64/1048576/100          1189 us         1189 us          594 bytes_per_second=840.822M/s null_percent=1 size=1048.58k
QuantileKernelInt64/1048576/10            889 us          889 us          786 bytes_per_second=1124.47M/s null_percent=10 size=1048.58k
QuantileKernelInt64/1048576/2             812 us          812 us          863 bytes_per_second=1.20332G/s null_percent=50 size=1048.58k
QuantileKernelInt64/1048576/1           0.601 us        0.601 us      1162030 bytes_per_second=1.58732T/s null_percent=100 size=1048.58k
QuantileKernelInt64/1048576/0             935 us          935 us          749 bytes_per_second=1069.31M/s null_percent=0 size=1048.58k

QuantileKernelDouble/1048576/10000       1116 us         1116 us          629 bytes_per_second=895.788M/s null_percent=0.01 size=1048.58k
QuantileKernelDouble/1048576/100         1222 us         1222 us          573 bytes_per_second=818.343M/s null_percent=1 size=1048.58k
QuantileKernelDouble/1048576/10           902 us          902 us          775 bytes_per_second=1108.27M/s null_percent=10 size=1048.58k
QuantileKernelDouble/1048576/2            878 us          878 us          797 bytes_per_second=1.1117G/s null_percent=50 size=1048.58k
QuantileKernelDouble/1048576/1          0.617 us        0.617 us      1134788 bytes_per_second=1.54598T/s null_percent=100 size=1048.58k
QuantileKernelDouble/1048576/0           1214 us         1214 us          577 bytes_per_second=823.777M/s null_percent=0 size=1048.58k

```

After
```
QuantileKernelInt32/1048576/10000         641 us          641 us         1091 bytes_per_second=1.52317G/s null_percent=0.01 size=1048.58k
QuantileKernelInt32/1048576/100           745 us          745 us          935 bytes_per_second=1.31037G/s null_percent=1 size=1048.58k
QuantileKernelInt32/1048576/10           1277 us         1277 us          549 bytes_per_second=783.305M/s null_percent=10 size=1048.58k
QuantileKernelInt32/1048576/2            1856 us         1856 us          377 bytes_per_second=538.889M/s null_percent=50 size=1048.58k
QuantileKernelInt32/1048576/1           0.627 us        0.627 us      1074304 bytes_per_second=1.52125T/s null_percent=100 size=1048.58k
QuantileKernelInt32/1048576/0             632 us          632 us         1109 bytes_per_second=1.54554G/s null_percent=0 size=1048.58k

QuantileKernelInt64/1048576/10000         370 us          370 us         1888 bytes_per_second=2.6384G/s null_percent=0.01 size=1048.58k
QuantileKernelInt64/1048576/100           414 us          414 us         1681 bytes_per_second=2.36152G/s null_percent=1 size=1048.58k
QuantileKernelInt64/1048576/10            686 us          686 us         1021 bytes_per_second=1.42442G/s null_percent=10 size=1048.58k
QuantileKernelInt64/1048576/2             906 us          906 us          774 bytes_per_second=1104.26M/s null_percent=50 size=1048.58k
QuantileKernelInt64/1048576/1           0.607 us        0.607 us      1155137 bytes_per_second=1.57133T/s null_percent=100 size=1048.58k
QuantileKernelInt64/1048576/0             364 us          364 us         1920 bytes_per_second=2.6835G/s null_percent=0 size=1048.58k

QuantileKernelDouble/1048576/10000       1115 us         1115 us          629 bytes_per_second=897.137M/s null_percent=0.01 size=1048.58k
QuantileKernelDouble/1048576/100         1244 us         1244 us          564 bytes_per_second=804.11M/s null_percent=1 size=1048.58k
QuantileKernelDouble/1048576/10           914 us          914 us          767 bytes_per_second=1093.93M/s null_percent=10 size=1048.58k
QuantileKernelDouble/1048576/2            900 us          900 us          778 bytes_per_second=1111.09M/s null_percent=50 size=1048.58k
QuantileKernelDouble/1048576/1          0.618 us        0.618 us      1079708 bytes_per_second=1.54343T/s null_percent=100 size=1048.58k
QuantileKernelDouble/1048576/0           1221 us         1221 us          573 bytes_per_second=819.062M/s null_percent=0 size=1048.58k
```

**Comment by cyb70289** [comment]
MinGW64 CI error is unrelated.
Travis CI passed in my personal repo https://travis-ci.org/github/cyb70289/arrow/builds/756695234

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 1]
```diff
@@ -239,6 +229,243 @@ struct QuantileExecutor {
...
+  using CType = typename InType::c_type;
+
+  void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    // cross point to benefit from histogram approach
+    // parameters estimated from ad-hoc benchmarks manually
+    static constexpr int kMinArraySize = 65536 * sizeof(int) / sizeof(CType);
+    static constexpr int kMaxValueRange = 65536;
+
+    const Datum& datum = batch[0];
+    if (datum.length() - datum.null_count() >= kMinArraySize) {
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/aggregate_quantile.cc:390]: Note that for int16/uint16, the max value range should always be satisfied.

**Review by pitrou — COMMENTED** [review]
Thank you very much. It may be useful to add a different benchmark size, or a benchmark with a larger value range?
```diff
diff --git a/cpp/src/arrow/compute/kernels/aggregate_benchmark.cc b/cpp/src/arrow/compute/kernels/aggregate_benchmark.cc
index 0454fa0ed..b1685d6a4 100644
--- a/cpp/src/arrow/compute/kernels/aggregate_benchmark.cc
+++ b/cpp/src/arrow/compute/kernels/aggregate_benchmark.cc
@@ -477,7 +477,7 @@ void QuantileKernelBench(benchmark::State& state) {
 }
 
 static void QuantileKernelBenchArgs(benchmark::internal::Benchmark* bench) {
-  BenchmarkSetArgsWithSizes(bench, {1 * 1024 * 1024});
+  BenchmarkSetArgsWithSizes(bench, {8 * 1024, 1 * 1024 * 1024});
 }
 
 #define QUANTILE_KERNEL_BENCHMARK(FuncName, Type)                                     \
```

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc [hunk 2]
```diff
@@ -239,6 +229,243 @@ struct QuantileExecutor {
...
+
+// histogram or sort approach per value range and size, only for integers
+template <typename InType>
+struct CountOrSortQuantiler {
+  using CType = typename InType::c_type;
+
+  void Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    // cross point to benefit from histogram approach
+    // parameters estimated from ad-hoc benchmarks manually
+    static constexpr int kMinArraySize = 65536 * sizeof(int) / sizeof(CType);
```
[the remark this answers is not in the document]
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:386]: Thanks for pointing this out. I confused memory size and item count. It should be 64K items, regardless of CType.
Benchmark result at https://docs.google.com/spreadsheets/d/1AmFrdLCIEaC4MaGXUDL5nJPmFuEvjDvsTqV0rDwn4GQ/edit?usp=sharing

File: cpp/src/arrow/compute/kernels/aggregate_quantile.cc (hunk 1 above)
Comment (cyb70289) [reply cpp/src/arrow/compute/kernels/aggregate_quantile.cc:390]: Yes. It looks still desirable to find min/max as it may not cover 64k range.

**Comment by cyb70289** [comment]
Benchmark updated to include integers with `wide` and `narrow` value range.
```
QuantileKernelInt32Wide/1048576/10000         1566 us         1566 us          447 bytes_per_second=638.552M/s null_percent=0.01 size=1048.58k
QuantileKernelInt32Wide/1048576/100           1906 us         1906 us          367 bytes_per_second=524.742M/s null_percent=1 size=1048.58k
QuantileKernelInt32Wide/1048576/10            2714 us         2714 us          258 bytes_per_second=368.515M/s null_percent=10 size=1048.58k
QuantileKernelInt32Wide/1048576/2             2431 us         2431 us          288 bytes_per_second=411.435M/s null_percent=50 size=1048.58k
QuantileKernelInt32Wide/1048576/1            0.627 us        0.627 us      1114130 bytes_per_second=1.52029T/s null_percent=100 size=1048.58k
QuantileKernelInt32Wide/1048576/0             1864 us         1864 us          376 bytes_per_second=536.539M/s null_percent=0 size=1048.58k
QuantileKernelInt32Narrow/1048576/10000        643 us          643 us         1088 bytes_per_second=1.51936G/s null_percent=0.01 size=1048.58k
QuantileKernelInt32Narrow/1048576/100          743 us          743 us          945 bytes_per_second=1.31356G/s null_percent=1 size=1048.58k
QuantileKernelInt32Narrow/1048576/10          1258 us         1258 us          557 bytes_per_second=795.072M/s null_percent=10 size=1048.58k
QuantileKernelInt32Narrow/1048576/2           1799 us         1799 us          387 bytes_per_second=555.725M/s null_percent=50 size=1048.58k
QuantileKernelInt32Narrow/1048576/1          0.644 us        0.644 us      1097860 bytes_per_second=1.47984T/s null_percent=100 size=1048.58k
QuantileKernelInt32Narrow/1048576/0            635 us          635 us         1105 bytes_per_second=1.53909G/s null_percent=0 size=1048.58k
QuantileKernelInt64Wide/1048576/10000          680 us          680 us         1034 bytes_per_second=1.43618G/s null_percent=0.01 size=1048.58k
QuantileKernelInt64Wide/1048576/100            905 us          905 us          772 bytes_per_second=1104.74M/s null_percent=1 size=1048.58k
QuantileKernelInt64Wide/1048576/10            1392 us         1392 us          505 bytes_per_second=718.177M/s null_percent=10 size=1048.58k
QuantileKernelInt64Wide/1048576/2              892 us          892 us          782 bytes_per_second=1120.52M/s null_percent=50 size=1048.58k
QuantileKernelInt64Wide/1048576/1            0.608 us        0.608 us      1138627 bytes_per_second=1.56729T/s null_percent=100 size=1048.58k
QuantileKernelInt64Wide/1048576/0              762 us          762 us          916 bytes_per_second=1.28095G/s null_percent=0 size=1048.58k
QuantileKernelInt64Narrow/1048576/10000        372 us          372 us         1887 bytes_per_second=2.62659G/s null_percent=0.01 size=1048.58k
QuantileKernelInt64Narrow/1048576/100          416 us          416 us         1687 bytes_per_second=2.34746G/s null_percent=1 size=1048.58k
QuantileKernelInt64Narrow/1048576/10           674 us          674 us         1040 bytes_per_second=1.44975G/s null_percent=10 size=1048.58k
QuantileKernelInt64Narrow/1048576/2            869 us          869 us          807 bytes_per_second=1.12415G/s null_percent=50 size=1048.58k
QuantileKernelInt64Narrow/1048576/1          0.609 us        0.609 us      1151528 bytes_per_second=1.56662T/s null_percent=100 size=1048.58k
QuantileKernelInt64Narrow/1048576/0            368 us          368 us         1912 bytes_per_second=2.65377G/s null_percent=0 size=1048.58k
QuantileKernelDouble/1048576/10000            1134 us         1134 us          614 bytes_per_second=882.09M/s null_percent=0.01 size=1048.58k
QuantileKernelDouble/1048576/100              1241 us         1241 us          565 bytes_per_second=805.54M/s null_percent=1 size=1048.58k
QuantileKernelDouble/1048576/10                924 us          924 us          747 bytes_per_second=1082.46M/s null_percent=10 size=1048.58k
QuantileKernelDouble/1048576/2                 889 us          889 us          791 bytes_per_second=1124.79M/s null_percent=50 size=1048.58k
QuantileKernelDouble/1048576/1               0.634 us        0.634 us      1127780 bytes_per_second=1.50509T/s null_percent=100 size=1048.58k
QuantileKernelDouble/1048576/0                1229 us         1229 us          570 bytes_per_second=813.899M/s null_percent=0 size=1048.58k
```

**Comment by cyb70289** [comment]
Travis CI passed: https://travis-ci.org/github/cyb70289/arrow/builds/757359819

**Review by pitrou — APPROVED** [review]
+1, thank you @cyb70289 !

**Pull request closed without merging**